external/llvm-project/mlir/lib/ExecutionEngine/RocmSystemDetect.cpp

```diff
@@ -0,0 +1,79 @@
+//===- RocmSystemDetect.cpp - Detect ROCm devices -------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements the system detection of ROCm devices on the current
+// system.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/ExecutionEngine/RocmSystemDetect.h"
+
+#include "llvm/Support/Error.h"
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wpedantic"
+#include "hip/hip_runtime.h"
+#pragma GCC diagnostic pop
+
+#define DEBUG_TYPE "execution-engine-rocm-system-detect"
+
+using namespace mlir;
+
+RocmSystemDetect::RocmSystemDetect() {
+  // collect all GPUs
+  int count = 0;
+  hipError_t herr = hipGetDeviceCount(&count);
+  if (herr != hipSuccess) {
+    llvm::errs() << "hipGetDeviceCount() should never fail\n";
+    return;
+  }
+
+  for (int i = 0; i < count; ++i) {
+    hipDeviceProp_t deviceProps;
+    herr = hipGetDeviceProperties(&deviceProps, i);
+    if (herr == hipSuccess) {
+      llvm::SmallString<32> chipName(deviceProps.gcnArchName);
+      llvm::StringRef chip = chipName;
+      llvm::StringRef vendor("AMD");
+      llvm::StringRef features;
+
+      auto idx = chipName.find_first_of(':');
+      if (idx != llvm::StringRef::npos) {
+        features = chipName.substr(idx + 1);
+        chip = chipName.substr(0, idx);
+      }
+      auto itr = std::find_if(begin(), end(), [&](const SystemDevice &dev) {
+        return dev.chip == chip;
+      });
```

Comment:
Yes, I was thinking we need to use a <chip>:<normalized_features> and have a special lookup for the case when a feature is not listed but matches when <feature>-. 

Question: can a feature with '-' be removed in normalization?

Reply:
Yeah, `-feature` means "assume feature is not a thing", `+feature` is "assume feature is a thing", and not mentioning feature means "it could go either way, make the conservative assumptions", last I checked

Reply:
For example, `-sramecc` will allow generating some code that `+sramecc` or no mention of sramecc wil not.